onnxruntime/core/providers/coreml/builders/impl/pad_op_builder.cc

```diff
@@ -0,0 +1,178 @@
+  GetShape(*input_defs[0], input_shape, logger);
+  const auto input_rank = SafeInt<int64_t>(input_shape.size());
+
+  const auto& pads_tensor = *model_builder.GetInitializerTensors().at(input_defs[1]->Name());            // pads
+  const auto& constant_value_tensor = *model_builder.GetInitializerTensors().at(input_defs[2]->Name());  // constant_value
+
+  float pad_value = 0.0f;
+  Initializer pad_value_raw_data_init(constant_value_tensor);
+  pad_value = pad_value_raw_data_init.DataAsSpan<float>()[0];
+  constant_padding_type->set_value(pad_value);
+
+  Initializer pads_initializer_raw_data(pads_tensor);
+  auto pads_span = pads_initializer_raw_data.DataAsSpan<int64_t>();
+
+  // Add padding
+  auto* height_border = coreml_pad->mutable_paddingamounts()->add_borderamounts();
+  height_border->set_startedgesize(pads_span[input_rank - 2]);
+  height_border->set_endedgesize(pads_span[2 * input_rank - 2]);
+  auto* width_border = coreml_pad->mutable_paddingamounts()->add_borderamounts();
+  width_border->set_startedgesize(pads_span[input_rank - 1]);
+  width_border->set_endedgesize(pads_span[2 * input_rank - 1]);
+
+  *layer->mutable_input()->Add() = input_defs[0]->Name();
+  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
+
+  model_builder.AddLayer(std::move(layer));
+
+  return Status::OK();
+}
+#endif
+
+// Operator support related
+
+bool PadOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                                     const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& initializers = input_params.graph_viewer.GetAllInitializedTensors();
+
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger))
+    return false;
+
+  if (input_shape.empty() || input_shape.size() < 2) {
+    LOGS(logger, VERBOSE) << "Pad requires input shape to be at least 2, input is "
+                          << input_shape.size() << "d shape";
+    return false;
+  }
+
+  if (std::find(input_shape.begin(), input_shape.end(), int64_t{0}) != input_shape.end()) {
+    LOGS(logger, VERBOSE) << "Pad input with zero elements for dimension is not supported";
+    return false;
+  }
+
+  {
+    NodeAttrHelper helper(node);
+    const auto mode = helper.Get("mode", "constant");
+    if (mode != "constant") {
+      LOGS(logger, VERBOSE) << "Only support constant mode Pad operator for now, mode: " << mode;
+      return false;
+    }
+
+    if (input_defs.size() < 3) {
+      LOGS(logger, VERBOSE) << "`constant_value` input is required for constant mode Pad op.";
+      return false;
+    }
+  }
+
+  // only support if `pads` input is known and does not contain negative values and only has non-zero padding values
+  // for last two dimensions.
+  {
+    const auto pads_initializer_it = initializers.find(input_defs[1]->Name());
+    if (pads_initializer_it == initializers.end()) {
+      LOGS(logger, VERBOSE) << "pads must be known";
```

Comment:
```suggestion
      LOGS(logger, VERBOSE) << "pads must be a constant initializer";
```

Technically we are just checking initializers but should really be also validating it's a constant initializer.

Ideally the error message should be meaningful to a user rather than a developer so that the user has a chance to correct things without creating a github issue. 'known' is potentially confusing as it doesn't convey that the value must be fixed and available when the model is being loaded.